consensus/src/twins_test.rs

```diff
@@ -200,3 +237,139 @@ fn basic_start_test() {
+        playground
+            .wait_for_messages(30, NetworkPlayground::votes_only)
+            .await;
+
+        // Check that the commit log for n0 is not empty
+        let mut commit_seen = false;
+        nodes[0].commit_cb_receiver.close();
+        if let Ok(Some(_node_commit)) = nodes[0].commit_cb_receiver.try_next() {
+            commit_seen = true;
+        }
+        assert!(commit_seen);
+
+        // Check that the commit log for n1 is empty
+        commit_seen = false;
+        nodes[1].commit_cb_receiver.close();
+        if let Ok(Some(_node_commit)) = nodes[1].commit_cb_receiver.try_next() {
+            commit_seen = true;
+        }
+        assert!(!commit_seen);
+    });
+}
+
+#[test]
+/// This test checks that the vote of a node and its twin
+/// should be counted as duplicate vote (because they have
+/// the same public keys)
+///
+/// Setup:
+///
+/// 4 honest nodes (n0, n1, n2, n3), and 1 twin (twin0)
+/// Create 2 partitions, p1=[n1, n3], p2=[n0, twin0, n2]
+///
+/// Test:
+///
+/// Extract enough votes to potentially form commits. Check
+/// that no node commits any block. This is because we need
+/// 3 nodes to form a quorum and no partition has enough votes
+/// (note there are 3 nodes in p2, but one of them is a twin,
+/// and its vote will be counted as duplicate of n0).
+///
+/// Run the test:
+/// cargo xtest -p consensus twins_vote_dedup_test -- --nocapture
+#[cfg(test)]
+fn twins_vote_dedup_test() {
+    let mut runtime = consensus_runtime();
+    let mut playground = NetworkPlayground::new(runtime.handle().clone());
+    let num_nodes = 4;
+    let num_twins = 1;
+    let (mut nodes, node_authors) = SMRNode::start_num_nodes_with_twins(
+        num_nodes,
+        num_twins,
+        &mut playground,
+        RotatingProposer,
+    );
+
+    // 4 honest nodes
+    let n0_twin_id = *playground.get_twin_ids(node_authors[0]).get(0).unwrap();
+    // twin of n0 has same author as node_authors[0]
+    let twin0_twin_id = *playground.get_twin_ids(node_authors[0]).get(1).unwrap();
+    let n1_twin_id = *playground.get_twin_ids(node_authors[1]).get(0).unwrap();
+    let n2_twin_id = *playground.get_twin_ids(node_authors[2]).get(0).unwrap();
+    let n3_twin_id = *playground.get_twin_ids(node_authors[3]).get(0).unwrap();
+
+    assert!(playground.split_network(
+        vec![n1_twin_id, n3_twin_id],
+        vec![twin0_twin_id, n0_twin_id, n2_twin_id],
+    ));
+
+    timed_block_on(&mut runtime, async {
+        playground
+            .wait_for_messages(1, NetworkPlayground::proposals_only)
+            .await;
+
+        // Pull enough votes to get a few commits.
+        // The proposer's votes are implicit and do not go in the queue.
+        playground
+            .wait_for_messages(30, NetworkPlayground::votes_only)
+            .await;
+
+        // No node should be able to commit because of the way partitions
+        // have been created
+
```

Comment:
Probably want to remove this line so the comment is more applicable to the block